go/consensus/tendermint/apps/keymanager/transactions.go

```diff
@@ -96,3 +98,116 @@ func (app *keymanagerApplication) updatePolicy(
 
 	return nil
 }
+
+func (app *keymanagerApplication) publishEphemeralSecret(
+	ctx *tmapi.Context,
+	state *keymanagerState.MutableState,
+	secret *api.SignedEncryptedEphemeralSecret,
+) error {
+	// Ensure that the runtime exists and is a key manager.
+	regState := registryState.NewMutableState(ctx.State())
+	rt, err := regState.Runtime(ctx, secret.Secret.ID)
+	if err != nil {
+		return err
+	}
+	if rt.Kind != registry.KindKeyManager {
+		return fmt.Errorf("keymanager: runtime is not a key manager: %s", secret.Secret.ID)
+	}
+
+	// Reject if the secret has been published.
+	_, err = state.EphemeralSecret(ctx, secret.Secret.ID, secret.Secret.Epoch)
+	switch err {
+	case nil:
+		return fmt.Errorf("keymanager: ephemeral secret for epoch %d already published", secret.Secret.Epoch)
+	case api.ErrNoSuchEphemeralSecret:
+		// Secret hasn't been published.
+	default:
+		return err
+	}
+
+	// Reject if the signer is not in the key manager committee.
+	signer := ctx.TxSigner()
+	kmStatus, err := state.Status(ctx, rt.ID)
+	if err != nil {
+		return err
+	}
+	inCommittee := false
+	for _, n := range kmStatus.Nodes {
+		if n == signer {
+			inCommittee = true
+			break
+		}
+	}
+	if !inCommittee {
+		return fmt.Errorf("keymanager: ephemeral secret can be published only by the key manager committee")
+	}
+
+	// Fetch RAK of the signer.
+	n, err := regState.Node(ctx, signer)
+	if err != nil {
+		return err
+	}
+	var kmrt *node.Runtime
+	for _, rt := range n.Runtimes {
+		if rt.ID == secret.Secret.ID {
+			// Skipping version check as key managers are running exactly one
+			// version of the runtime.
+			kmrt = rt
+			break
+		}
+	}
+	if kmrt == nil {
+		return fmt.Errorf("keymanager: signer is not a key manager node")
+	}
+	tee := kmrt.Capabilities.TEE
+
+	var rak signature.PublicKey
+	switch {
+	case tee == nil:
+		rak = api.InsecureRAK
+	case tee.Hardware == node.TEEHardwareInvalid:
+		rak = api.InsecureRAK
```

Comment:
Yes. Will add comment to the code also.

Reply:
Turns out that's not okay as we allow a node to re-register with `tee=nil` after it is added to the key manager committee. In that case the node could use insecure RAK to publish an ephemeral secret even though the committee is running on SGX hardware.

Reply:
This check is probably incorrect:
https://github.com/oasisprotocol/oasis-core/blob/c7d180e360786316dc895041b0a31c4bce6e42a0/go/registry/api/api.go#L798-L801

It should be changed so that in case it is `nil` but `regRt.TEEHardware` indicates it shouldn't be, registration should fail.

Reply:
Agree, will fix.